File: week10_sort/350.intersection-of-two-arrays-ii.cpp
#include "common.h"

class Solution {
public:
  vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
    const int len1 = nums1.size();
    const int len2 = nums2.size();
    if (len1 == 0 || len2 == 0) return {};
    sort(nums1.begin(), nums1.end());
    sort(nums2.begin(), nums2.end());
    vector<int> res;
    int i = 0, j = 0;
    while (i < len1 && j < len2) {
      if (nums1[i] == nums2[j]) {
        res.push_back(nums1[i]);
        ++i;
        ++j;
      } else if (nums1[i] < nums2[j]) ++i;
      else ++j;
    }
    return res;
  }
};

int main() {
  vector<int> nums1, nums2;
  INPUT_ARRAY(nums1);
  INPUT_ARRAY(nums2);
  PRINT_ARRAY(Solution().intersect(nums1, nums2));
  return 0;
}

File: week10_sort/350-intersection-of-two-arrays-ii.rs
pub struct Solution {}

impl Solution {
    pub fn intersect(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        let mut nums1 = nums1;
        let mut nums2 = nums2;
        nums1.sort();
        nums2.sort();
        let mut result = Vec::new();
        let (mut i, mut j) = (0, 0);
        while i < nums1.len() && j < nums2.len() {
            if nums1[i] == nums2[j] {
                result.push(nums1[i]);
                i += 1; j += 1;
            } else if nums1[i] < nums2[j] {
                i += 1;
            } else {
                j += 1;
            }
        }
        result
    }
}

fn main() {
    let out = Solution::intersect(vec![1, 2, 2, 1], vec![2, 2]);
    assert_eq!(out, [2, 2]);

    let out = Solution::intersect(vec![4, 9, 5], vec![9, 4, 9, 8, 4]);
    assert_eq!(out, [4, 9]);
}
